src/components/OntologyExplorer/index.tsx

```diff
@@ -131,22 +136,65 @@ export default function OntologyExplorer({ graph }: OntologyExplorerProps): JSX.
 
     Side effect: sets the render & simulation state.
     */
+
     if (dagState) {
-      const { nodes, links } = dagState;
+      const { nodes, links } = dagState;      
+
+      const nodeToHullRoot = new Map();
+      let flag = true;
+      let height = 7;
+      const allHullRoots: string[] = [];
+      while (flag && height >= 2) {
+        flag=false;
+        const hullRoots = [...heightMap].filter(([_, v]) => v === height ).map(([k,_])=>k) // get all hullRoots of a certain height
+        hullRoots.forEach((item)=>{ // for each root
+          const hullNodes = getHullNodes(item, ontology, nodes);
+          hullNodes.forEach((n: any)=>{ // for each node in root
+            if (!nodeToHullRoot.has(n.id)) { // if not already assigned a root
+              if (!allHullRoots.includes(item)) {
+                allHullRoots.push(item); // add root
+              }
+              nodeToHullRoot.set(n.id,item); // set the root.
+            }
+          })
+        })
+        try{
+          nodes.forEach((node)=>{ // for each node,
+            const id = node.id;
+            if (!nodeToHullRoot.has(id)) { // check if node doesn't have a root
+              flag=true; // set true
+              throw(Error);
+            }
+          })
+        } catch {}
+        height-=1;
+      }
+      const hullToNodes = new Map();
+      [...nodeToHullRoot].forEach((kv)=>{
```

Comment:
see above comments about slamming maps into arrays just to iterate over them

Couple of thoughts (verging on opinions):
1. in ES6/TS, for-of is the simplest way to write iterators.  The entire language (e.g. Map/Set) have support for it via the iteration protocol.
2. for-of also avoids anti-patterns like above where you create copies of the entire object just to iterate over it (or the empty catch to terminate a forEach).

I realize the functional model has its benefits. So this is really a question of clarity given the current language state.

Reply:
The map --> array --> .forEach is just me not knowing any better 🤡. Good to know that maps can be iterated over with for-of!